tensorflow/compiler/xla/service/gpu/ir_emitter_nested.cc

```diff
@@ -295,53 +305,64 @@ void EmitAMDGPUAtomicAdd(llvm::IRBuilder<>* builder,
-    llvm::VectorType* half2type = llvm::VectorType::get(builder->getHalfTy(), 
-																						llvm::ElementCount::getFixed(2));
+ 
+  bool is_shared_mem = output_address_type->getPointerAddressSpace() == 3;
+  const auto& debug_options =
+      ir_emitter_context.hlo_module().config().debug_options();
+  auto compute_capability = ir_emitter_context.rocm_compute_capability();
+
+  if ((element_type == F32 || element_type == F64) && !is_shared_mem &&
+      // MI100/MI200 atomics flush denormals by default for fp32
+      (compute_capability.gfx9_mi300_or_later() ||
+       debug_options.xla_gpu_ftz() || element_type != F32)) {
+    llvm_ir::EmitCallToIntrinsic(
+        llvm::Intrinsic::amdgcn_global_atomic_fadd, {output_ptr, source},
+        {source->getType(), output_ptr->getType(), source->getType()}, builder);
+    return;
+  }
+
+  if ((element_type == F16 || element_type == BF16) && !is_shared_mem &&
+      paired_atomics) {
+    llvm::VectorType* half2type = llvm::VectorType::get(
+        source->getType(), llvm::ElementCount::getFixed(2));
     auto i16 = builder->getInt16Ty();
     auto i32 = builder->getInt32Ty();
     auto i64 = builder->getInt64Ty();
     auto half2ptr = llvm::PointerType::get(half2type, 1);
     auto intptr = builder->CreatePtrToInt(output_address, i64);
-    auto alignment = builder->CreateAnd(intptr, llvm::ConstantInt::get(i64, 2ull));
+    auto alignment =
+        builder->CreateAnd(intptr, llvm::ConstantInt::get(i64, 2ull));
     intptr = builder->CreateAnd(intptr, llvm::ConstantInt::get(i64, ~3ull));
     output_ptr = builder->CreateIntToPtr(intptr, half2ptr);
 
     auto shift = builder->CreateShl(builder->CreateTrunc(alignment, i32), 3);
     auto i16src = builder->CreateBitCast(source, i16);
     auto intsrc = builder->CreateZExt(i16src, i32);
     source = builder->CreateShl(intsrc, shift);
-    source = builder->CreateBitCast(source, half2type); 
-
-    llvm::Module* module = builder->GetInsertBlock()->getModule();
-    std::vector<llvm::Type*> ir_input_types{half2ptr, half2type};
-
-    llvm::FunctionType* callee_type = llvm::FunctionType::get(
-      half2type, ir_input_types, false);
-
-    // Declares the callee if it is not declared already.
-    llvm::Function* callee = llvm::dyn_cast<llvm::Function>(
-        builder->GetInsertBlock()
-            ->getModule()
-            ->getOrInsertFunction("llvm.amdgcn.global.atomic.fadd.v2f16.p1.v2f16", callee_type)
-            .getCallee());
+    source = builder->CreateBitCast(source, half2type);
 
-    callee->addFnAttr(llvm::Attribute::NoUnwind);
-    callee->setMemoryEffects(llvm::MemoryEffects::argMemOnly());
-
-    builder->CreateCall(callee, {output_ptr, source});//llvm_ir::AsArrayRef(operands));
+    llvm_ir::EmitCallToIntrinsic(
+        element_type == BF16 ? llvm::Intrinsic::amdgcn_global_atomic_fadd_v2bf16
+                             : llvm::Intrinsic::amdgcn_global_atomic_fadd,
```

Comment:
Do you mean `llvm::Intrinsic::amdgcn_global_atomic_fadd_v2fp16`?

Reply:
No, no. It is like it is.

Reply:
is `llvm::Intrinsic::amdgcn_global_atomic_fadd` for fp32? what about fp16? I'm wondering where can I see this intrinsic?

Reply:
Thanks for the explanation! if I want to check related `llvm::Intrinsic::amdgcn`, I'm wondering where should I look for? is somewhere here? https://github.com/llvm/llvm-project/tree/main/llvm/lib/Target/AMDGPU

Reply:
It is generated from this file https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/IntrinsicsAMDGPU.td#L2906